### PR TITLE
Add unit tests for LDAP auth provider

### PR DESCRIPTION
**Pull request by PennyScissors**
## Issue: 

https://github.com/rancher/rancher/issues/42324

## Problem

Need unit tests for LDAP auth provider. 
 
## Solution

Initially the idea was to write a basic unit test for the AuthenticateUser function as we identified it to be a good start point, but after exploring multiple approaches this was not possible for several reasons. One being a higher level interface all providers belong to prevents me from modifying the function signature (to inject mocks) without significant refactoring and the codebase structure in general is very limiting because many of our functions have multiple responsibilities. Instead, I implemented unit tests for the inner functions that make up the important logic within AuthenticateUser.

Some minor refactoring was necessary in order to mock some structures successfully.
1. Functions taking in an ldap connection struct were updated to take in an ldap client interface instead to allow for mocking the ldap connection calls.
2. The Connect function that initiates the ldap connection was moved outside of the loginUser function as it was easier to do this and pass in the connection than trying to mock the function call.
3. The getLDAPConfig function now expects a UnstructureClient argument. This was necessary to be able to mock CRUD calls (in this case a Get call inside the function). The UnstructuredClient is an interface whereas the ObjectClient is not, so the cleaner approach was to use the UnstructuredClient as the parameter.

## Engineering Testing
Steps I followed to test this changes in Rancher v2.7-head f308df3a69e763951f16ef8d5934393571ddea10
1. Enable FreeIPA auth provider
2. Log out of Rancher
3. Log in using a FreeIPA user

## QA Testing Considerations
See Regressions Considerations below

## Regressions Considerations
The changes I made passed a basic test of enabling FreeIPA and logging in as a FreeIPA user, but a change like switching from a struct pointer parameter to an interface could potentially introduce edge cases. This should be further tested by QA to make sure all the basic functionality provided by LDAP based auth providers works as intended.

### CODE REVIEW

**Comment by crobby** [comment]
Just got around to taking a first (super quick) peek now.  Maybe add a wantError: true case for ldap_client_test.go?

**Comment by PennyScissors** [comment]
> Just got around to taking a first (super quick) peek now. Maybe add a wantError: true case for ldap_client_test.go?

I added 2 more cases, another one for invalid credentials but this one mocks an error response from the ldap server, and one for a user without access permissions

**Comment by crobby** [comment]
For Test_ldapProvider_getLDAPConfig are there any cases you can think of where we'd expect an error?

File: pkg/auth/providers/ldap/ldap_provider_test.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+	"github.com/stretchr/testify/assert"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var (
```
Comment (crobby) [inline pkg/auth/providers/ldap/ldap_provider_test.go:24]: Should these be const instead of var?

Comment (PennyScissors) [reply pkg/auth/providers/ldap/ldap_provider_test.go:24]: I will change this in the next commit

**Comment by PennyScissors** [comment]
> For Test_ldapProvider_getLDAPConfig are there any cases you can think of where we'd expect an error?

> Ideally, we have some tests for each of the exported functions. Any thoughts on how tricky it might be to get tests for at least the most commonly used exported functions (AuthenticateUser, SearcPrincipals, GetPrincipal)?

Discussed these in a huddle and agreed there isn't any negative cases to test at the moment except potentially one with the common.Decode function, but since it is in the common package and is not LDAP specific, it is lower priority and can be addressed later. As for testing the exported functions, this is something we cannot do at the moment due to the higher level AuthProvider interface preventing us from modifying function signatures to allow for mock injection. However, as a short term solution the inner functions (like getLDAPConfig) that make up AuthenticateUser and some of those other exported functions are being tested individually to get past this problem.

**Review by crobby — COMMENTED** [review]
I think we'll want to squash the 2 commits before merging, but I'm good with what's here.  Certainly a step forward.

File: pkg/auth/providers/common/ldap/ldap_util.go [hunk 2]
```diff
@@ -170,7 +170,7 @@ func GetGroupSearchAttributesForLDAP(ObjectClass string, config *v3.LdapConfig)
 	return groupSeachAttributes
 }
 
-func AuthenticateServiceAccountUser(serviceAccountPassword string, serviceAccountUsername string, defaultLoginDomain string, lConn *ldapv3.Conn) error {
+func AuthenticateServiceAccountUser(serviceAccountPassword string, serviceAccountUsername string, defaultLoginDomain string, lConn ldapv3.Client) error {
```
Comment (nflynt) [inline pkg/auth/providers/common/ldap/ldap_util.go:173]: This may be my own ignorance of the language at work here, but passing in the `ldapv3.Client` by **value** looks strange. What happens if we need to authenticate two service accounts in a row, and we pass the same client (from the call site) in twice? Does the state get synchronized in some way that I'm missing?

Comment (crobby) [inline pkg/auth/providers/common/ldap/ldap_util.go:173]: Is this getting called anywhere in the tests?  Not seeing it....if it's not being called, do we need to change the param?

File: pkg/auth/providers/ldap/ldap_provider.go [hunk 3]
```diff
@@ -111,18 +112,32 @@ func (p *ldapProvider) TransformToAuthProvider(authConfig map[string]interface{}
...
+func toBasicLogin(input interface{}) (*v32.BasicLogin, error) {
 	login, ok := input.(*v32.BasicLogin)
 	if !ok {
-		return v3.Principal{}, nil, "", errors.New("unexpected input type")
+		return nil, errors.New("unexpected input type")
+	}
+	return login, nil
+}
+
+func (p *ldapProvider) AuthenticateUser(ctx context.Context, input interface{}) (v3.Principal, []v3.Principal, string, error) {
```
Comment (nflynt) [inline pkg/auth/providers/ldap/ldap_provider.go:123]: While we're in here, since `AuthenticateUser` is exported it needs a docstring. (This is a new rule being applied to old code, but we should follow it for any function we change during a refactor like this.)

File: pkg/auth/providers/common/ldap/ldap_util.go (hunk 2 above)
Comment (nflynt) [reply pkg/auth/providers/common/ldap/ldap_util.go:173]: d'oh, that's an _interface_. Ignore me, this is fine.

File: pkg/auth/providers/ldap/ldap_client_test.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+				userMGR:               tt.fields.userMGR,
+				tokenMGR:              tt.fields.tokenMGR,
+				certs:                 tt.fields.certs,
+				caPool:                tt.fields.caPool,
+				providerName:          tt.fields.providerName,
+				testAndApplyInputType: tt.fields.testAndApplyInputType,
+				userScope:             tt.fields.userScope,
+				groupScope:            tt.fields.groupScope,
+			}
+			got, got1, err := p.loginUser(tt.args.lConn, tt.args.credential, tt.args.config, tt.args.caPool)
```
Comment (nflynt) [inline pkg/auth/providers/ldap/ldap_client_test.go:180]: nit: the loginUser call sites calls `got` -> **userPrincipal**, and `got1` -> **groupPrincipals**. I would use those names here, and use **wantUser** and **wantGroups** in the test table, just for clarity

**Review by nflynt — APPROVED** [review]
I have some tiny nits regarding naming, but the code changes look good to me and the tests are readable and easy enough to follow. This seems like a good starting point for us to build on :+1:

**Review by crobby — DISMISSED** [review]

**Review by nflynt — APPROVED** [review]
